### PR TITLE
feat(): captcha to mitigate abuse, closes #9

### LINKED ISSUE #9: Captcha
Require solving a captcha to register a new user.

### PR DESCRIPTION
**Pull request by nils-wisiol**
Introduce captcha to prevent abuse while maintaining easy usability.

### CODE REVIEW

File: api/desecapi/pdns.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                "records": [
+                    {
+                        "type": type,
+                        "name": name,
+                        "disabled": False,
+                        "content": value,
+                    }
+                ],
+                "ttl": 60,
```
Comment (peterthomassen) [inline api/desecapi/pdns.py]: maybe make this an argument to this function, with a default value of 60

File: api/desecapi/pdns.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+def _pdns_get(url):
+    r = requests.get(settings.POWERDNS_API + url, headers=headers)
+    if (r.status_code < 200 or r.status_code >= 300) and r.status_code != 404:
+        raise Exception(r.text)
+    return r
+
+
+def _delete_or_replace(name, type, value):
+    """
+    Return pdns API json to either replace or delete a record, depending on value is empty or not.
```
Comment (peterthomassen) [inline api/desecapi/pdns.py]: This json replaces or deletes a record _set_. If you have 5 A records and then send the replacement json to pdns, you'll be left with only one record.

File: api/desecapi/pdns.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    else:
+        return \
+            {
+                "changetype": "DELETE",
+                "type": type,
+                "name": name
+            }
+
+
+def create_native_zone(name):
```
Comment (peterthomassen) [inline api/desecapi/pdns.py]: rename to `create_zone` with 2nd argument `kind='NATIVE'`? Later, once pdns knows how to do DNSSEC via API, we can switch the default to `MASTER`.

File: api/desecapi/templates/emails/captcha/subject.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+deSEC dedyn.io Account Suspended
```
Comment (peterthomassen) [inline api/desecapi/templates/emails/captcha/subject.txt]: Add domain name

File: api/desecapi/tests/testdyndns12update.py [hunk 5]
```diff
@@ -121,4 +122,28 @@ def testManualIPv6(self):
...
+
+        httpretty.reset()
+        httpretty.enable()
+        httpretty.HTTPretty.allow_net_connect = False
+
+        domain = self.owner.domains.all()[0]
+        domain.arecord = '10.1.1.1'
+        domain.save()
+
+        httpretty.register_uri(httpretty.POST, settings.POWERDNS_API + '/zones')
```
Comment (peterthomassen) [inline api/desecapi/tests/testdyndns12update.py]: the `POST` line is unnecessary

File: api/desecapi/tests/testdyndns12update.py [hunk 6]
```diff
@@ -121,4 +122,28 @@ def testManualIPv6(self):
...
+        self.assertIP(ipv4='127.0.0.1')
+
+    def testSuspendedUpdates(self):
+        self.owner.captcha_required = True
+        self.owner.save()
+
+        httpretty.reset()
+        httpretty.enable()
+        httpretty.HTTPretty.allow_net_connect = False
+
```
Comment (peterthomassen) [inline api/desecapi/tests/testdyndns12update.py]: add a test of domain creation in locked state

File: api/desecapi/tests/testregistration.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+from desecapi import models
+from datetime import timedelta
+from django.utils import timezone
+from django.core import mail
+from desecapi.emails import send_account_lock_email
+from desecapi import settings
+
+
+class RegistrationTest(APITestCase):
+    def testRegistrationSuccessful(self):
```
Comment (peterthomassen) [inline api/desecapi/tests/testregistration.py]: change function names from camel case to underscore style

File: api/desecapi/views.py [hunk 8]
```diff
@@ -267,3 +276,51 @@ def sendDonationEmails(donation):
...
+
+
+def unlock(request, email):
+    # if this is a POST request we need to process the form data
+    if request.method == 'POST':
+        # create a form instance and populate it with data from the request:
+        form = UnlockForm(request.POST)
+        # check whether it's valid:
+        if form.is_valid():
+            user = None
```
Comment (peterthomassen) [inline api/desecapi/views.py]: delete this line

**Comment by nils-wisiol** [comment]
still missing: limit domain number

**Comment by nils-wisiol** [comment]
add closed issues to commit msgs

File: api/desecapi/pdns.py (hunk 2 above)
Comment (nils-wisiol) [inline api/desecapi/pdns.py]: I changed the function name to `_delete_or_replace_rrset` and updated the doc comment

File: api/desecapi/pdns.py (hunk 1 above)
Comment (nils-wisiol) [inline api/desecapi/pdns.py]: changed as requested

File: api/desecapi/pdns.py (hunk 3 above)
Comment (nils-wisiol) [inline api/desecapi/pdns.py]: changed as requested

File: api/desecapi/templates/emails/captcha/subject.txt (hunk 4 above)
Comment (nils-wisiol) [inline api/desecapi/templates/emails/captcha/subject.txt]: will be using the 'first' domain name of the user at hand, and 'deSEC dedyn.io' if no domain name given.

File: api/desecapi/tests/testdyndns12update.py (hunk 6 above)
Comment (nils-wisiol) [inline api/desecapi/tests/testdyndns12update.py]: added another test

File: api/desecapi/tests/testdyndns12update.py (hunk 5 above)
Comment (nils-wisiol) [inline api/desecapi/tests/testdyndns12update.py]: removed

File: api/desecapi/tests/testregistration.py (hunk 7 above)
Comment (nils-wisiol) [inline api/desecapi/tests/testregistration.py]: done

File: api/desecapi/views.py (hunk 8 above)
Comment (nils-wisiol) [inline api/desecapi/views.py]: done

**Comment by nils-wisiol** [comment]
ready to merge?

File: api/desecapi/templates/emails/captcha/content.txt [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  {{url}}
+
+and unlock your account for future use. Until unlocked, your
+account will be kept in read-only mode, that is, you won't be able
+to register any domains or update IP addresses (or any other
+records). Accounts that are not unlocked are subject to deletion
+after two weeks.
+
+If you need mass registrations for a legit purpose, please
+contact us directly. Just reply this email, we will be happy to
```
Comment (peterthomassen) [inline api/desecapi/templates/emails/captcha/content.txt]: reply _to_ this email

File: api/desecapi/templates/emails/captcha/content.txt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+get in touch with you.
+
+As we use Google Recaptcha for our CAPTCHA, we recommend using
+your browsers anonymous mode to solve it.
```
Comment (peterthomassen) [inline api/desecapi/templates/emails/captcha/content.txt]: your _browser's_ anonymous

File: api/desecapi/templates/emails/captcha/content.txt (hunk 9 above)
Comment (nils-wisiol) [inline api/desecapi/templates/emails/captcha/content.txt]: fixed

File: api/desecapi/templates/emails/captcha/content.txt (hunk 10 above)
Comment (nils-wisiol) [inline api/desecapi/templates/emails/captcha/content.txt]: fixed

File: api/desecapi/views.py [hunk 11]
```diff
@@ -34,6 +51,11 @@ def perform_create(self, serializer):
             ex.status_code = 409
             raise ex
 
+        if self.request.user.limit_domains is not None and self.request.user.domains.count() == self.request.user.limit_domains:
```
Comment (nils-wisiol) [inline api/desecapi/views.py]: fixed

**Comment by peterthomassen** [comment]
@nils-wisiol, feel free to merge/deploy. My connection here is a bit shaky, so I don't dare.

**Pull request merged**